components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -272,133 +314,496 @@ do_registration_ends ()
+  const auto p2d       = m_p2d;
 
-    const auto& ph = phys.get_header();
-    const auto& dh = dyn.get_header();
+  const auto field_loop = KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const int i = team.league_rank();
 
-    const auto& phys_alloc_prop = ph.get_alloc_properties();
-    const auto& dyn_alloc_prop  = dh.get_alloc_properties();
+    switch (phys_layout(i)) {
+      case 0:
+      case 1:
+      case 2:
+      {
+        if (pack_alloc_property(i) == 0) {
+          set_dyn_to_zero<pack_type>(team);
+          team.team_barrier();
 
-    const bool phys_pack_alloc = phys_alloc_prop.template is_compatible<pack_type>();
-    const bool dyn_pack_alloc  = dyn_alloc_prop.template  is_compatible<pack_type>();
-    const bool phys_small_pack_alloc = phys_alloc_prop.template is_compatible<small_pack_type>();
-    const bool dyn_small_pack_alloc  = dyn_alloc_prop.template  is_compatible<small_pack_type>();
+          local_remap_fwd_2d<pack_type>(team, num_cols, lid2elgp, p2d);
+        } else if (pack_alloc_property(i) == 1) {
+          set_dyn_to_zero<small_pack_type>(team);
+          team.team_barrier();
 
-    const int itl = m_is_state_field[i] ? tl.np1 : -1;
+          local_remap_fwd_2d<small_pack_type>(team, num_cols, lid2elgp, p2d);
+        } else {
+          set_dyn_to_zero<Real>(team);
+          team.team_barrier();
 
-    const LayoutType lt = get_layout_type(ph.get_identifier().get_layout().tags());
-    switch (lt) {
-      case LayoutType::Scalar2D:
-      case LayoutType::Vector2D:
-      case LayoutType::Tensor2D:
-        remap_bwd_2d(phys,dyn,itl);
+          local_remap_fwd_2d<Real>(team, num_cols, lid2elgp, p2d);
+        }
         break;
-      case LayoutType::Scalar3D:
-      case LayoutType::Vector3D:
-      case LayoutType::Tensor3D:
-        if (phys_pack_alloc && dyn_pack_alloc) {
-          remap_bwd_3d_impl<pack_type>(phys,dyn,itl);
-        } else if (phys_small_pack_alloc && dyn_small_pack_alloc) {
-          remap_bwd_3d_impl<small_pack_type>(phys,dyn,itl);
+      }
+      case 3:
+      case 4:
+      case 5:
+      {
+        if (pack_alloc_property(i) == 0) {
+          set_dyn_to_zero<pack_type>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<pack_type>(team, num_cols, lid2elgp, p2d);
+        } else if (pack_alloc_property(i) == 1) {
+          set_dyn_to_zero<small_pack_type>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<small_pack_type>(team, num_cols, lid2elgp, p2d);
         } else {
-          remap_bwd_3d_impl<Real>(phys,dyn,itl);
+          set_dyn_to_zero<Real>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<Real>(team, num_cols, lid2elgp, p2d);
         }
         break;
+      }
       default:
         ekat::error::runtime_abort("Error! Unhandled case in switch statement.\n");
     }
-  }
+  };
+  const auto policy = ekat::ExeSpaceUtils<KT::ExeSpace>::get_default_team_policy(num_fields, 1);
```

Comment:
So, this is something worth checking going fwd, but I _think_ it might be best to ||ize over `num_fields*num_cols`. The fields might be unbalanced: if qsize>>1, remapping Q is much more work than remapping, say, T.

For 2d quantities, it leaves just 1 iteration in the local remap loop (which is a waste of threads), but 3d quantities can fully occupy 4 warps (at 128 levels). OTOH, I think surface pressure might be the only 2d field to remap (and I'm not 100% sure we need to remap it either), so it's ok to have some idle threads on 1 remapped fields, if we gain something on the 3d fields...

This can be adjusted in a second PR though.

Reply:
One side benefit of pulling the `icol` loop at the top level is that you only need switches to extract subviews, while the inner loop (for 3d fields, at the k level, that is) is the same, since the TTR would always happen on 1d views (you can subview elgp[0], time level (if any), elgp[1], and elgp[2]). That might skim a few lines of repeated code.

Reply:
Yeah, that makes sense if some field are much larger. I'll work on the implementation after this PR.